launcher/minecraft/mod/tasks/LocalResourcePackParseTask.cpp

```diff
@@ -178,6 +179,168 @@ bool processZIP(ResourcePack& pack, ProcessingLevel level)
+    };
+
+    auto it = text_to_hex.find(text_color);
+
+    if (it == text_to_hex.end()) {
+        qWarning() << "Invalid color string in component!";
+        result = "#000000"; // return black if color not found
+        return false;
+    }
+
+    result = it.value();
+    return true; 
+}
+
+bool getBoolOrFromText(const QJsonValue& val, bool& result) 
+{
+    if (val.isUndefined()) {
+        result = false;
+        return true;
+    }
+
+    if (val.isBool()) {
+        result = val.toBool();
+        return true;
+    } else if (val.isString()) {
+        auto bool_str = val.toString(); 
+
+        if (bool_str == "true") {
+            result = true;
+            return true;
+        } else if (bool_str == "false") {
+            result = false;
+            return true;
+        } else {
+            qWarning() << "Invalid bool value in component!";
+            return false;
+        }
+    } else {
+        qWarning() << "Invalid type where bool expected!";
+        return false;
+    }
+}
+
+bool readFormat(const QJsonObject& obj, TextFormat& format)
+{
+    auto text = obj.value("text");
+    auto color = obj.value("color");
+    auto bold = obj.value("bold");
+    auto italic = obj.value("italic");
+    auto underlined = obj.value("underlined");
+    auto strikethrough = obj.value("strikethrough");
+    auto extra = obj.value("extra");
+
+    if (color.isString()) {
+        // colors can either be a hex code or one of a few text colors
+        auto col_str = color.toString();
+
+        const QRegularExpression hex_expression("^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})$");
+        const auto hex_match = hex_expression.match(col_str);
+
+        if (hex_match.hasMatch()) {
+            format.color = color.toString();
+        } else {
+            if (!textColorToHexColor(color.toString(), format.color)) {
+                qWarning() << "Invalid color type in component!";
+                return false;
+            }
+        }
+    }
+
+    return getBoolOrFromText(bold, format.bold) && getBoolOrFromText(italic, format.italic) &&
+           getBoolOrFromText(underlined, format.underlined) && getBoolOrFromText(strikethrough, format.strikethrough);
+}
+
+void appendBeginFormat(TextFormat& format, QString& toAppend)
+{
+    toAppend.append("<font color=\"" + format.color + "\">");
+    if (format.bold)
+        toAppend.append("<b>");
+    if (format.italic)
+        toAppend.append("<i>");
+    if (format.underlined)
+        toAppend.append("<u>");
+    if (format.strikethrough)
+        toAppend.append("<s>");
+}
+
+void appendEndFormat(TextFormat& format, QString& toAppend)
+{
+    toAppend.append("</font>");
+    if (format.bold)
+        toAppend.append("</b>");
+    if (format.italic)
+        toAppend.append("</i>");
+    if (format.underlined)
+        toAppend.append("</u>");
+    if (format.strikethrough)
+        toAppend.append("</s>");
```

Comment:
If you use multiple tags the end tags should be in reverse order of the beginning so instead of getting:
```<font color=\"#FFAA00\"><b><i>SomeText</font></b></i>```
you would get: 
```<font color=\"#FFAA00\"><b><i>SomeText</i></b></font>```
That is a more correct in HTML